### PR TITLE
Delete file on rollback in HiveRecordWriter

### PR DESCRIPTION
**Pull request by losipiuk**
Currently if table write operation was interrupted by user
the files which were already partly written to table directory were not
deleted.
In RecordFileWriter.rollback() the RecordWriter.close() was called with
abort parameter set to true. But implementations of RecordWriters ignore
that parameter.
That resulted with garbage files for CTAS/INSERT flows which did not use
temporary directory (when S3 FS was in use).

By explicitly deleting file in RecordFileWriter.rollback() we make this
proble less probable to surface.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java [hunk 1]
```diff
@@ -161,6 +163,8 @@ public void rollback()
     {
         try {
             recordWriter.close(true);
+            // perform explicit delete of written file here as recordWriter.close() ignore the abort flag.
+            path.getFileSystem(conf).delete(path, false);
```
Comment (pnowojski) [inline presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java]: What will happen if the created file is not yet visible? I wasn't able to find an answer to this question here:

http://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html#ConsistencyModel

but it states that: `A process writes a new object to Amazon S3 and immediately lists keys within its bucket. Until the change is fully propagated, the object might not appear in the list. ` However there is no explanation what happens with "delete-after-write" :/

Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java]: I don't know. When I tested that it seems to behave correctly - the files are deleted - even though delete is issed immediatly after closing file outputstream. But there may be some corner cases.

File: presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java [hunk 2]
```diff
@@ -77,6 +78,7 @@ public RecordFileWriter(
             TypeManager typeManager)
     {
         this.path = requireNonNull(path, "path is null");
+        this.conf = requireNonNull(conf, "conf can not be null");
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java]: nit: `conf is null`

File: presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java [hunk 3]
```diff
@@ -161,6 +163,8 @@ public void rollback()
     {
         try {
             recordWriter.close(true);
+            // perform explicit delete of written file here as recordWriter.close() ignore the abort flag.
+            path.getFileSystem(conf).delete(path, false);
         }
         catch (IOException e) {
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java]: What if `close()` succeeds and `delete()` throws `IOException`, do we still want to throw `HIVE_WRITER_CLOSE_ERROR`?

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java]: I think throwing an exception here is wrong, since `rollback()` is called in a loop in `HivePageSink.doAbort()` to rollback all the writers. This means we can leak file descriptors or other resources.

File: presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java (hunk 1 above)
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java]: Grammar nit: ignore -> ignores

**Comment by electrum** [comment]
Please make the same change in `RcFileFileWriter`

File: presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java (hunk 3 above)
Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java]: But `recordWriter.close()` can throw exception anywat, so if there is issue with lost resources, we have the problem already  in the current codebase.

Or do you suggest to remove the throwing of exception from `catch` clause in `rollback` at all?

**Comment by losipiuk** [comment]
> Please make the same change in RcFileFileWriter

I would skip that unless @dain confirms that he has not code for that alrady in flight.
It is marked as todo and it seems Dain is currently working on that.

**Comment by losipiuk** [comment]
added grammar fixups

**Comment by electrum** [comment]
I think we should remove the throw and replace it with log warning. I do believe the existing code can leak.

**Comment by losipiuk** [comment]
> I think we should remove the throw and replace it with log warning. I do believe the existing code can leak.

I added commit which replaces `throw` witl `log.warn`.

Yet it seems to me that it would be better to change the loop which calls rollback in `HivePageSink.doAbort()` to accumulate potenial exceptions thrown by `rollback` calls.
And in case any of the `rollback` calls threw an exception, rethrow a composite one.

This way we do not implement internal of `rollback` method to match way how it is called much deeper on the stack.
And we handle both `RecordFileWriter` and `RcFileWriter` with one logic.

What do you think?

**Comment by electrum** [comment]
The new RC writer is all checked in and ready to use. The todo was a "this should be fixed" rather than "work in progress".

**Comment by electrum** [comment]
Doing it in HivePageSink sounds good. I like that approach.

**Comment by nezihyigitbasi** [comment]
I guess the same problem also applies to the `commit()` method, throwing in that method may result in a leak.

**Comment by losipiuk** [comment]
Upadated patch.
Added file deleting to `RcFileWriter` and code which prevents breaking the loop of `rollback` calls if one of theses throws an exception.

I also changed original commit to perform delete in `finally` section even if `RecordWriter.close` throws an exception.

Per @nezihyigitbasi comment:

> I guess the same problem also applies to the commit() method, throwing in that method may result in a leak.

Isn't it mitigated by the fact that in case of exception on `commit()` rollback loop would be called afterwards? At least it is what I expect.

**Comment by nezihyigitbasi** [comment]
@losipiuk You are probably right. I see rollback being called if commit doesn't go through in `TableWriterOperator::close()`.

**Comment by losipiuk** [comment]
@nezihyigitbasi, @electrum please take a look when you have a chance :)

File: presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java [hunk 4]
```diff
@@ -160,7 +162,13 @@ public void commit()
     public void rollback()
     {
         try {
-            recordWriter.close(true);
+            try {
+                recordWriter.close(true);
+            }
+            finally {
+                // perform explicit delete of written file here as recordWriter.close() ignores the abort flag.
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java]: perform explicit deletion here as implementations of RecordWriter.close() often ignore the abort flag.

File: presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriterFactory.java [hunk 5]
```diff
@@ -139,8 +140,14 @@ else if (ColumnarSerDe.class.getName().equals(storageFormat.getSerDe())) {
                 });
             }
 
+            Callable<Void> outputFileDeleteRoutine = () -> {
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriterFactory.java]: why don't you make this runnable and get rid of the return?

Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriterFactory.java]: you can also put the try catch here and remove it from `RcFileFileWriter.java` where you ignore it.

File: presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriter.java [hunk 6]
```diff
@@ -46,12 +47,14 @@
...
 {
     private final RcFileWriter rcFileWriter;
+    private final Callable<Void> outputFileDeleteRoutine;
     private final int[] fileInputColumnIndexes;
     private final List<Block> nullBlocks;
     private final Optional<Supplier<RcFileDataSource>> validationInputFactory;
 
     public RcFileFileWriter(
             OutputStream outputStream,
+            Callable<Void> outputFileDeleteRoutine,
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriter.java]: how about renaming this to `rollbackAction` or sth similar?

File: presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriter.java [hunk 7]
```diff
@@ -136,10 +145,14 @@ public void commit()
...
-            rcFileWriter.close();
-            // todo delete file
+            try {
+                rcFileWriter.close();
+            }
+            finally {
+                outputFileDeleteRoutine.call();
+            }
         }
-        catch (IOException e) {
```
Comment (nezihyigitbasi) [inline presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriter.java]: why did you change this?

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java [hunk 8]
```diff
@@ -212,12 +216,24 @@ public void abort()
...
+                catch (Exception e) {
+                    exceptions.put(writer, e);
+                }
             }
         }
+        if (!exceptions.isEmpty()) {
+            exceptions.entrySet().forEach(
+                    entry -> log.warn("exception '%s' while rollback on %s", entry.getValue(), entry.getKey())
+            );
+            throw new PrestoException(HIVE_WRITER_CLOSE_ERROR, "Error rolling back write to Hive");
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java]: Put one of the exceptions as the cause for this. It would be nice to have the file name for that one as well.

@dain, @electrum, and I considered putting all of them in suppressed, and decided not to that. But we feel like that is not likely going to give more information. And when we add a list to suppressed, we have to be careful about not having too many of them, which is additional hassle.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java [hunk 9]
```diff
@@ -212,12 +216,24 @@ public void abort()
...
+                    writer.rollback();
+                }
+                catch (Exception e) {
+                    exceptions.put(writer, e);
+                }
             }
         }
+        if (!exceptions.isEmpty()) {
+            exceptions.entrySet().forEach(
+                    entry -> log.warn("exception '%s' while rollback on %s", entry.getValue(), entry.getKey())
```
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java]: Why not print the logs as you encounter that? The list doesn't appear to be necessary.

File: presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriterFactory.java (hunk 5 above)
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriterFactory.java]: @losipiuk 
I had a discussion with @nezihyigitbasi. We agreed that having try...catch where you put it is better. Please ignore the last comment about moving the try catch.

**Comment by losipiuk** [comment]
@haozhun back to you :)

**Comment by losipiuk** [comment]
@haozhun  Any update on this one? It is small so let's get rid of it :)

File: presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriterFactory.java (hunk 5 above)
Comment (haozhun) [inline presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriterFactory.java]: `Callable` looks better.

**Comment by haozhun** [comment]
Merged #7468

**Comment by losipiuk** [comment]
Thanks
